### ISSUE TITLE
Browsing slow or not responding with server timeouts

### ISSUE THREAD

**Issue by ExSport** [UniversalMediaServer/UniversalMediaServer#708]
I installed latest snapshot "Build: 5fe999673 (2015-10-25)" and tested to play few children videos and pictures from mobile for my parents. I spotted (not first time) that renderer (Panasonic TV) totally disconnects and I had to exit DLNA on TV and return back after a while to get any response.
In log I can see browsing sends list with 20 files in bulk but UMS have to check all files in folder before any list is sent to renderer. I should expect that UMS will read all files without parsing them, then takes first bulk (depending on ordering set), parse them and send them to TV. And in the background it continues with the rest in folder.
Due to the behavior of parsing all files before anything is sent to renderer, it takes to UMS more then renderer can wait and it will timeout and renderer disconnects. My tesing folder had only 200 photos but parsing all of them at time is time consuming as any file is checked for compatibility, parsed with MediaInfo, tsMuxeR engine checking against pictures and other irrelevant steps which takes easily minute or more :-(
I think using multithreading and cutting job to more separate bulks will greatly help in renderer response and removing timeouts.
@Nadahar, @SubJunk Somewhere I read you talked about the network code and I support your opinion to recheck the network code as I am sure there are some bugs as many times when parents play some short clips or pictures continuously, PS3 and PanTV shows disconnections after some time what breaks whole playlist/slideshow. Sometimes playing TS files which are streamed as is without any modification, first e.g. 15 files play normally and later at the end of next files renderer as PS3 shows disconnections etc. so it seems something is lost in network communication. Sometines PS3 continues on other files in list, other times it stops or completely hangs.
Ehm, sorry for OT, will try to find original post about network code (Netty etc...)
Thx

**Comment by Nadahar** [comment]
@ExSport Do you mean to say that any of these problems are new with the recent snapshot? If so, what's the last version you know of that this didn't happen?

**Comment by ExSport** [comment]
@Nadahar
"Not the first time"  in  context means that it doesn't started with latest version but that it still happens with latest version.
Unfortunately I am limited in testing as UMS is used by parents and I get home hardly once a month so I have limited time to fix their complaints what means trying to update to the latest build and sometimes making some logs, checking code and sometimes providing the fix. Unfortunately I don't have any smart tv so outside I am limited to Android phone and UMS:(

**Comment by SubJunk** [comment]
@ExSport I agree about multithreading. Unfortunately I don't know much about it

**Comment by Nadahar** [comment]
I don't know the code in question well enough to know how this is handled by UMS, but from what I've seen so far calls to external programs like FFMpeg and MEncoder is done in separate threads where the calling thread is waiting for the other thread to finish and return the results before continuing. I'm not sure if MediaInfo is handled the same way - but in either case it's not really a parallel execution because they are called sequentially. So, the next thread aren't started before the previous has finished. If UMS is queried by more than one renderer at once, it would happen in parallel though. It would be possible to just fire a lot of threads and then organize the collection of the results, but it would require a complete rewriting of the logic, and could possibly be quite tricky to get right. In any case, multithreading itself doesn't solve everything, there are multiple issues. You have to look at the flow of data, because multithreading involves synchronization of memory and locks before any data can be exchanged between threads. That means that multithreading is only really beneficial if there are a lot of heavy calculations that can be done as a "separate operation" without needing to exchange information with the rest of the application, like generating hashes/checksums. I doubt it's much of this type of processing going on here. Secondly

**Comment by ExSport** [comment]
Right now I expect only what is needed regarding sort order when files are sent to renderer to display is the name. Question is if it is possible to cut whole file list to smaller parts which renderer asks before all parsing jobs are done on all files what is really time consuming.
So e.g. read all 400 filenames from folder and instead of doing all parsing tasks for all files at a time before anything is sent,  that UMS will read all filenames,  prettify them and then will parse other relevant info e.g. with MediaInfo for bulk of files renderer asks e.g. at 20 pieces and rest in background so UMS can predict number of files to display and send them as they are parsed. We are talking about first bulk of files as renderer generally asks for first 20? items and later asks for the rest. Most important is sending first 20, for the rest UMS has much more time to parse them.
Btw. when talking about DB, few months ago I reported found bug that UMS parses new files,  stores them in DB but then instead of reading them again from DB files are parsed again and again until UMS is restarted.After that parsed info is correctly read from DB.
Only my two cents. I know many guys who lost attitude when renderer returned timeouts with UMS so they replaced it with an alternative because they never waited to parse all data for bigger folders or tried few restarts to see UMS started to work for them without timeouts or disconnections.
Easily reproducible with fotos or videos from e.g. camcorder = lots of smaller files in one folder.
In this example it seems to the user that UMS disconnects or connection timeouts and when user tries to browse another folder or playing any file, everything get stuck. But in the background, hidden to the user, UMS intensively parses all files so all other renderer requests are ignored until UMS will finish his previous job, parsing all xyz files.
Any thoughts how to make things better? :))

**Comment by Nadahar** [comment]
@ExSport As stated already, I don't know this in detail. The way I have interpreted the pieces of the code regaring this the I've seen, is that that UMS needs to check each entry for that specific renderer to decide if that item should be presented or not, and of what type it should be. Not all entries should be presented to every renderer, and what type should be presented could also vary. From this I conclude that it's not possible to do it any other way than to let UMS build the complete folder until the first chunk is presented.

It that's a good way to do it or not is another question, but I'm sure changing it would do a lot of the "magic" that's done adapting everything to each renderer will become impossible.

Many things in UMS seem to me to have been adapted for a certain kind of use in mind (watching movies) and other parts, like watching images or listening to audio seems to have been neglected/forgotten about when decisions have been made.

**Comment by SubJunk** [comment]
@Nadahar yeah that is true and I've been guilty of having a video-centric approach. If we ever manage to write good automated tests that could be largely safeguarded.

**Comment by Nadahar** [comment]
Just for the record, I don't mean to say this is fortunate but I don't believe there's an easy fix to this. It would probably require rewriting the entire browse logic, loosing renderer adaptability as a consequence. I still think it should be looked at to see if it's possible to optimize the process without sacrificing too much.

I have my hands on too much already, so I'm not going to look at that now. I also feel that there's other people in this team that has much more knowledge than me about the DLNA tree building process.

**Comment by ExSport** [comment]
I didn't say it will be easy or doable:-)
Btw. it seems to me UMS serializes requests (in FIFO order) so when one huge folder is parsed in the background so renderer gets timeout and user will try to browse another almost empty folder, from user side it seems all requests are revoked as all new browse actions are ignored until huge folder parsing job is finished. In other words until first job is finished, all other renderer requests are not satisfied. I think user will expect that this huge parsing will go in the background so browsing another folder with few files will be parsed and shown before background job have to be finised so it will not block all other user actions. Can't test it right now but I think when another user with another renderer will browse folder at that time, renderer will receive response as requests are ordered in another FIFO queue so one renderer job will not block all others but only itself but as I said I can't test it right now and only deduct it.
Is it possible somehow optimize this behavior? Some advanced multithreading :-)
If it will be possible to not block other requests then browsing some huge folder UMS can predict how long will parsing took for whole folder so after e.g. 5secs UMS will send some fake file with name like "Parsing 800 files, return after a minute or so...". In that case renderer will not timeout, users will know that something is hapenning and they can continue with other folder (if this parsing will not revoke any other actions done later as I described above)
It is only my own fabulation but want to start some discussion if someone else recommends some other way how to do it more clever, more easily, without some huge recoding etc. :+1: 
Any thoughts @UniversalMediaServer/developers ?

**Comment by Nadahar** [comment]
@ExSport I think have to be a member of the development team for @UniversalMediaServer/developers to work (a bit strange to me, but I guess it makes sense in some scenarios). Their attention will be called now that I've used it though.

I think you are wrong about the FIFO. In fact, renderers that invoke multiple requests fast create some havoc in UMS as I see it, firing e.g. multiple browse processes running in parallel. The request handler spawns a thread for each incoming request that "lives its own life". Again, I think IO is the bottleneck that probably explains why it seems to be FIFO. As long as that big browse job is running, getting disk time will be difficult and the small browse job will be starved as a result. Whenever they try to use the same objects (folders, items) they will have to wait for object synchronization as well. Synchronization calls are FIFO as far as I know, but that's completely beyond our control (it's controlled by the OS).

Again, I haven't studied the code in enough detail to be sure, but this is my impression of how it works. I bet there's optimalizations that could have been done though, but this is a complicated and critical part of the code that nobody really want to fiddle too much with. One thing I've notices is that there's way too many calls to ```File.exists()``` (in my view), I don't think those that wrote this realized that each of these generate IO activity. Even though it doesn't take much bandwidth, it could require the heads to move quite a lot, slowing down total IO performance. I don't know how much impact this has though, and it probably has to do with both the HDD's caching policy and the HDD speed. There's also quite a lot of calls to external programs to determine transcoding, which I guess is using a big part of the time. This is done for every item. It's a bit harder to imagine how this could be done more effectively.

@UniversalMediaServer/developers @SubJunk Please feel free to fill in or correct me.

**Comment by ExSport** [comment]
I will correct what I meant by FIFO. I know there are limited processes running in the background so there is some multithreading in place. I wanted to say that it seems requests done at time UMS is busy with other tasks, new ones are orderer at the end of pipe so until slow and long parsing of some huge folder is done, no other request will be finished before previous parsing is done.
With my english skills it is hard to say in one sentence so shortly. Parsing two folders, one with 500 files, other one with 5 files. Browsing first one will start huge parsing job which will lead to timeout on renderer. So user will try to access second folder. If UMS will start parsing these 5files, I would expect renderer will not timeout but after some longer while it will list those 5 files, before all 500 files must be parsed from first folder. Unfortunately from user side it seems first folder parsing terminated whole UMS server as files in folder are not listed and trying to list another folder do the same, timeouts, no response from UMS so user will give up and close DLNA browser or restarts UMS.
I know parsing is quite time consuming process but I am sure it can be done better to minimize timeouts when user browse folders on renderers. I hope so :+1:  
I will try to make some huge fake folder and one small folder and will test if all renderer requests will timeout afterwards or the only one which requested browsing of this huge folder.
Unfortunately will not be at home for next two weeks :-(

**Comment by Nadahar** [comment]
@ExSport I know all about the english challenges, I'm not native either (I'm Norwegian) and often feel I don't manage to explain myself clearly. I did understand what you meant here though (FIFO is hard to misunderstand), and my answer to that is that it is fully multithreaded. There's a separate "dispatcher" thread that only receives incoming messages, and spawns a thread to handle each specific request and then passes the request to that thread. All requests are handles in parallel.

But, as I tried to say, there are many other things in play here. First there is the IO queue which isn't controlled by UMS. I don't know what exact rules apply there, and it probably varies with OS and hardware, but my usual experience with disk IO is that it's shared. With that I mean that for each new request, every request goes slower - they all share the total IO. However, it's a numbers game and if one thread spawn a lot of requests (like the one browsing a big folder), the time slots assigned to the few requests from the small folder browsing thread will be very small. It's not shared on a per thread basis, but per IO request. That said, I think each thread works relatively sequential, so the big thread probably won't spawn a lot of parallel requests, but a long sequence of requests. 

In addition to this there is memory synchronization and object locking. This is stricly FIFO as far as I know, and that goes for any objects both of the browsing threads will need to access. It could potentially be a problem here, that the first browsing thread locks an object that the other thread needs, and hold that lock long for a long time. That will effectively make the second thread wait for the first. There are a lot of shortcomings in UMS's locking strategy (or the lack of), and such a scenario is entirely possible. It's very complicated to get a full overview over how locking works though, and it's also very difficult to debug because debugging is thousands of times slower then running the code and it completely alters the timing between the thread being debugged and the one running in it's own pace. That's why a consistent strategy is needed, this is very hard to debug, but unfortunately it's already somewhat of a disaster in this area - and Java with it's "lazy synchronization" keyword that generally locks way more than needed makes it much worse than if it'd been done manually.

If you want to test where the problem lies, I'd recommend creating the folders on a SSD. That will dramaticly reduce IO access time and eliminate head movent waiting, so that could help indicate whether IO access or thread locking is the problem - or if it's at true that requests are being processes in a serialized way. As I don't use SSD's because of their low reliability I can't test that.
